### ISSUE TITLE
glob exclude does not have effect on Android resource merging

### ISSUE THREAD

**Issue by menny** [bazelbuild/bazel#873]
In this example app (https://github.com/menny/BazelPlayground/tree/glob_resources_bug), I have a resource file `src/main/res/xml/config_something.xml` that I do not want to include in the app's resources, and instead, I want to use a different one from another folder `src/debug/res/xml/config_something.xml`.

So, I wrote this `android_binary` rule"

```
android_binary(
        name = "app",
        srcs = glob(["src/main/java/**/*.java"]),
        resource_files = glob(include=["src/main/res/**"], exclude=["src/main/res/xml/config_something.xml"]) +
                        glob(include=["src/debug/res/xml/config_something.xml"]),
        custom_package = "net.evendanan.bazelplayground",
        manifest = "src/main/AndroidManifest.xml",
        deps = ["//javalib",
                "@android_support_v4//:lib"],
        visibility = ["//visibility:public"]
```

Building the app result in an error:

```
➜  BazelPlayground git:(glob_resources_bug) /Users/menny/dev/menny/bazel/output/bazel build app
INFO: (02-09 17:03:34.498) Found 1 target...
ERROR: (02-09 17:03:35.722) /Users/menny/dev/menny/BazelPlayground/app/BUILD:1:1: Processing resources failed: resources_processor failed: error executing command bazel-out/host/bin/external/bazel_tools/tools/android/resources_processor --aapt bazel-out/host/bin/external/androidsdk/aapt_binary --annotationJar external/androidsdk/tools/support/annotations.jar ... (remaining 17 argument(s) skipped): com.google.devtools.build.lib.shell.BadExitStatusException: Process exited with status 1.
Feb 09, 2016 5:03:35 PM com.google.devtools.build.android.AndroidResourceProcessingAction main
SEVERE: Error during merging resources
/var/folders/6x/c5p2dxf13075rcdf4plh6x0c0000gn/T/android_resources_tmp5203993952957251087/tmp-deduplicated/app/src/debug/res/xml/config_something.xml: Error: Duplicate resources: /var/folders/6x/c5p2dxf13075rcdf4plh6x0c0000gn/T/android_resources_tmp5203993952957251087/tmp-deduplicated/app/src/debug/res/xml/config_something.xml:xml/config_something, /var/folders/6x/c5p2dxf13075rcdf4plh6x0c0000gn/T/android_resources_tmp5203993952957251087/tmp-deduplicated/app/src/main/res/xml/config_something.xml:xml/config_something
    at com.android.ide.common.res2.DataSet.checkItems(DataSet.java:380)
    at com.android.ide.common.res2.DataSet.loadFromFiles(DataSet.java:239)
    at com.android.ide.common.res2.ResourceSet.loadFromFiles(ResourceSet.java:47)
    at com.google.devtools.build.android.AndroidResourceProcessor.mergeData(AndroidResourceProcessor.java:439)
    at com.google.devtools.build.android.AndroidResourceProcessingAction.main(AndroidResourceProcessingAction.java:321)

Target //app:app failed to build
```

You'll notice that the error says that we have duplicate resources:  
1. /var/folders/6x/c5p2dxf13075rcdf4plh6x0c0000gn/T/android_resources_tmp5203993952957251087/tmp-deduplicated/app/src/**debug**/res/xml/config_something.xml:xml/config_something
2. /var/folders/6x/c5p2dxf13075rcdf4plh6x0c0000gn/T/android_resources_tmp5203993952957251087/tmp-deduplicated/app/src/**main**/res/xml/config_something.xml:xml/config_something

Using the latest _HEAD_ from _master_.

For context, my project should support Bazel and Gradle at the same time, and we are using Gradle's resource override feature to use a different resource for `debug` builds.

**Comment by ahumesky** [comment]
I suspect this is another sandboxing problem. On linux this works:
bazel build app:app

but this fails with the same error above:
bazel build app:app --spawn_strategy=standalone

**Comment by menny** [comment]
Interesting.
Do you think that a similar fix can be applied in this case?

**Comment by ahumesky** [comment]
So the problem here is that the exclude attr is effectively being ignored because the android tools are operating on directories, rather than explicitly listed files. One way around this is to move the excluded files into their own directory, say src/release/res/.... and src/debug/res/.... and keep the common resources in src/main/res/.... and then create different targets using the different resources:
```python
RELEASES = {
  "": glob(["src/release/res/**/*.xml"]),
  "_debug": glob(["src/debug/res/**/*.xml"]),
}

[android_binary(
        name = "app" + suffix,
        srcs = glob(["src/main/java/**/*.java"]),
        resource_files = glob(include=["src/main/res/**"]) + resources,
        custom_package = "net.evendanan.bazelplayground",
        manifest = "src/main/AndroidManifest.xml",
        deps = [
          "//javalib",
          "@android_support_v4//:lib",
          "@android_appcompat_v7//:lib",
          "@android_support_design//:lib",
          "//libs/play-services-base-8.4.0:lib"],
        visibility = ["//visibility:public"]
) for suffix, resources in RELEASES.items()]
```

I'm not sure if this will work with your need to maintain bazel and gradle at the same time though.

**Comment by menny** [comment]
@ahumesky, you mentioned that this setup build successfully on Linux. How was this problem resolved there?

**Comment by ahumesky** [comment]
bazel has action sandboxing on linux, so when the resource processing action runs in its sandbox, it can only see the files that were explicitly put into the sandbox, which are the ones from the resource_files glob, even though in the real filesystem there might be more files in those directories. Mac doesn't have a sandbox (yet?), so the android resource tools saw all the files in the directories.

**Comment by ahumesky** [comment]
On linux, if you turn off the sandbox with --spawn_strategy=standalone, then you see the same error

**Comment by damienmg** [comment]
@ahumesky Do we still need to do something here?

**Comment by ahumesky** [comment]
As far as I know we're still waiting for sandboxing on mac. However, there's a new resource merger that's being worked on that might not be vulnerable to this.

**Comment by polycoder** [comment]
👍 on the need for a fix for this globbing issue, or a resource merger that can handle merging conflicts more like gradle.

**Comment by ahumesky** [comment]
We've made significant progress on the new merger, but we're still working on a small number of corner cases before we turn it on. Should be on the order of weeks though before we can make this available.

**Comment by philwo** [comment]
@ahumesky Can you give an update on this? Curious if we can close it as fixed now.

**Comment by ahumesky** [comment]
Now that we have sandboxing on the mac, we should give this a try to check if this works. I don't have a mac at hand at the moment, but can try it maybe later today or Monday.

**Comment by aj-michael** [comment]
This was fixed (at least to the extent that it will be fixed) when I rolled out the new merger in 61c3b1caddf747268940e6ebd1afc4cf33543002 last November, sorry for the delay in closing. Now if you try to build the sample you get a CONFLICT message, but the merger resolves it and the build succeeds:

```
$ bazel build app --spawn_strategy=standalone
INFO: Found 1 target...
INFO: From Processing Android resources for //app:app:
Apr 11, 2017 4:17:28 PM com.google.devtools.build.android.AndroidDataMerger doMerge
WARNING: 
CONFLICT: xml/config_something is provided with ambiguous priority from:
        app/src/main/res/xml/config_something.xml
        app/src/debug/res/xml/config_something.xml
Target //app:app up-to-date:
  bazel-bin/app/app_deploy.jar
  bazel-bin/app/app_unsigned.apk
  bazel-bin/app/app.apk
INFO: Elapsed time: 19.072s, Critical Path: 16.73s
```
